From b520c7a5f354942f2f5479d1f0ed2b85ca79c27c Mon Sep 17 00:00:00 2001
From: Nadav Hugim <nhugim@habana.ai>
Date: Thu, 16 Dec 2021 09:27:13 +0200
Subject: [PATCH] [SW-66652] gaudi: security: fix fake otp writing

we need to save the hash of pub key in memory so we can
read it after bootint (only in fake security)

Change-Id: Ie263ae0b95565336ab9325c0c05171a1e7f6ce5d
Signed-off-by: Nadav Hugim <nhugim@habana.ai>
---

diff --git a/arch/gaudi/arm/security.c b/arch/gaudi/arm/security.c
index e4d24c2..24f520c 100644
--- a/arch/gaudi/arm/security.c
+++ b/arch/gaudi/arm/security.c
@@ -166,9 +166,25 @@
 	boot_info_set(HL_BOOT_INFO_SECURITY_FUSE, efuse, sizeof(*efuse));
 }
 
-static void save_hash_pub_key(u8 *ppub_key, int key_length)
+static void save_hash_pub_key(const void *img, size_t img_size)
 {
-	sha256_csum(ppub_key, key_length, hash_value);
+	const void *fit_sec_off;
+	struct hl_fit_sec_desc fit_sec;
+
+	/* hl fit files are padded so the size is 4 bytes aligned */
+	img_size = ROUND_UP4(img_size);
+
+	/* security descriptor is at the footer of the image */
+	fit_sec_off = (const void *)((u8 *)img + img_size);
+
+	memcpy(&fit_sec, fit_sec_off, FIT_SEC_DESC_SIZE);
+
+	if (sizeof(struct hl_fit_sec_desc) > fit_sec.desc_size) {
+		preboot_puts("FAKE_SECURITY FAIL to save pub_key\n");
+		return;
+	}
+
+	sha256_csum(fit_sec.pub_key, FIT_PUB_KEY_SIZE, hash_value);
 
 	/*
 	 * Pass public key value to u-boot.
@@ -176,11 +192,10 @@
 	 * used in fake security mode.
 	 */
 	preboot_security_sp_sram_write_key_hash((u8 *)hash_value, SHA256_SUM_LEN);
-
 	preboot_puts("\nFAKE_SECURITY is ON and pub_key saved\n");
 }
 #else
-static void save_hash_pub_key(u8 *ppub_key, int key_length) {}
+static void save_hash_pub_key(const void *img, size_t img_size) {}
 
 static void preboot_security_copy_otp(struct hl_sec_otp *otp)
 {
@@ -2179,9 +2194,10 @@
 	const void *fit_sec_off;
 	const void *pub_key_fit;
 
-	if (!preboot_security_is_enabled())
+	if (!preboot_security_is_enabled()) {
+		save_hash_pub_key(img, img_size);
 		return 0;
-
+	}
 	/* hl fit files are padded so the size is 4 bytes aligned */
 	img_size = ROUND_UP4(img_size);
 
@@ -2254,9 +2270,6 @@
 	}
 
 	*key_idx = fit_sec.key_num;
-
-	save_hash_pub_key(fit_sec.pub_key, FIT_PUB_KEY_SIZE);
-
 	return 0;
 }
 
